doc/developer/design/20240625_subsource_deprecation_blue_green_sources.md

```diff
@@ -0,0 +1,317 @@
+- https://github.com/MaterializeInc/materialize/issues/24843
+- https://github.com/MaterializeInc/materialize/issues/15897
+- https://github.com/MaterializeInc/materialize/issues/17021
+
+## The Problems
+
+TLDR:
+- Subsources are confusing and not implemented uniformly across source types
+- It's impossible to do a blue-green workflow in sources for handling
+  schema-changes in upstream systems
+
+Users currently have to learn about *sources*, *subsources*, *tables*,
+*materialized views*, *indexes*, and *sinks* when onboarding to Materialize.
+
+If they'd like to ingest data from both an upstream Postgres/MySQL database
+and a Kafka topic, they also need to learn the fragmented user interface
+for dealing with data ingested from upstream data sources. Our existing
+model of a *source* does not provide a consistent interface among source types.
+
+Single-output Kafka sources put their data under the source's primary relation,
+whereas multi-output Postgres/MySQL sources do not allow querying the source's primary
+relation and instead contain *subsources*, an entirely separate concept.
+
+Our existing *source* models also make it difficult to deal with upstream schema changes,
+such that users need to drop and recreate a *source* or *subsource* and any dependent objects
+to begin ingesting data with a new schema. This can cause long-periods of downtime if the
+new source needs to reingest the entire upstream history. This can also cause them to
+lose their _reclocking_ information which makes it impossible to implement a correct
+end-to-end *source* -> *sink* pipeline.
+
+This change aims to simplify the user experience and create a more unified
+model for managing data ingested from an upstream source, while providing
+more flexibility in handling upstream schema changes.
+
+
+## Success Criteria
+
+- The mental overhead & learning time for users to ingest data from an upstream system
+  is decreased.
+- All source types use a consistent interface to retrieve progress information and
+  ingested data. The primary *source* object represents the same concept across
+  all source types.
+- The same upstream table/topic can be ingested into multiple collections in Materialize,
+  each with a potentially different projection of the upstream schema. This enables
+  a user to do blue-green schema-change workflows without unavailability.
+- *sources* can maintain their progress collection while the set of collections being
+  ingested are modified, allowing an end-to-end *source* -> *sink* schema-change workflow
+  that does not emit redundant data (reusable reclocking).
+
+## Out of Scope
+
+<!--
+What does a solution to this problem not need to address in order to be
+successful?
+
+It's important to be clear about what parts of a problem we won't be solving
+and why. This leads to crisper designs, and it aids in focusing the reviewer.
+-->
+
+- Accomplishing 'subsource demuxing' using the concept of *subsources* in-use today.
```

Comment:
I think we should close it once we accomplish its intended goals, rather than at the design phase. That should happen after Step 6 in my implementation plan below, which gets us to demuxing into multiple tables before doing the work to refactor the primary source collection and progress collection interfaces and support for tables with  kafka sources, which I think will be the trickiest parts of this work overall.

Reply:
Agree—that's my preference as well!